src/commands/auth.ts

```diff
@@ -78,84 +78,145 @@ const preserveCredentials = async (
   });
 };
 
+type TokenSetContents = {
+  user_id: string;
+} & TokenSet;
+
+const isValidTokenSet = (tokenSet: TokenSet): boolean => {
+  // Only require refresh_token for refresh attempts
+  return !!tokenSet.refresh_token;
+};
+
+const isCompleteTokenSet = (
+  tokenSet: TokenSet,
+): tokenSet is Required<TokenSet> => {
+  return !!(
+    tokenSet.access_token &&
+    tokenSet.refresh_token &&
+    tokenSet.expires_at
+  );
+};
+
+const handleExistingToken = async (
+  tokenSet: TokenSet,
+  props: AuthProps,
+  credentialsPath: string,
+): Promise<{ apiKey: string; apiClient: Api<unknown> } | null> => {
+  // Validate token set
+  if (!isValidTokenSet(tokenSet)) {
+    log.debug('Invalid token set found');
+    return null;
+  }
+
+  // Check if token needs refresh or is missing
+  if (tokenSet.expired() || !tokenSet.access_token) {
```

Comment:
I think we could reduce nesting and make this function a bit easier to follow if we instead check:
```
if (!!tokenSet.access_token && !tokenSet.expired()) {
  ... // return existing access token
}
... // try to refresh
```